[-Wdeprecated-copy-with-dtor] definition of implicit copy assignment operator is deprecated

With the latest clang update it seems that MIOpen is hit with the following warning as error in build process:

```
# clang --version
AMD clang version 16.0.0 (ssh://gerritgit/lightning/ec/llvm-project amd-mainline-open 23126 df03e317430e013f05093fef7112cb245b25dd1c)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm/llvm/bin
```

at least the following files need to be updated:
![image](https://user-images.githubusercontent.com/7841904/229428723-0291b78c-8c41-4c77-a846-4884257aaf05.png)

For example, the warning messages look like:
```
/MIOpen/src/include/miopen/performance_config.hpp:43:13: warning: definition of implicit copy assignment operator for 'PerfConfig' is deprecated because it has a user-declared destructor [-Wdeprecated-copy-with-dtor]
    virtual ~PerfConfig() = default;
            ^
/MIOpen/src/include/miopen/performance_config.hpp:53:8: note: in implicit copy assignment operator for 'miopen::solver::PerfConfig' first required here
struct PerfConfigBase : PerfConfig
       ^
/MIOpen/src/include/miopen/solver.hpp:270:8: note: in implicit copy assignment operator for 'miopen::solver::PerfConfigBase<miopen::solver::PerformanceConfigConvAsm3x3U>' first required here
struct PerformanceConfigConvAsm3x3U : PerfConfigBase<PerformanceConfigConvAsm3x3U>
       ^
/MIOpen/src/include/miopen/serializable.hpp:109:14: note: in implicit copy assignment operator for 'miopen::solver::PerformanceConfigConvAsm3x3U' first required here
        self = out;
             ^
/MIOpen/src/include/miopen/performance_config.hpp:62:26: note: in instantiation of function template specialization 'miopen::solver::SerDes<>::Deserialize<miopen::solver::PerformanceConfigConvAsm3x3U>' requested here
        return SerDes<>::Deserialize(static_cast<Derived&>(*this), s);
                         ^
/MIOpen/src/include/miopen/solver.hpp:277:38: note: in instantiation of member function 'miopen::solver::PerfConfigBase<miopen::solver::PerformanceConfigConvAsm3x3U>::Deserialize' requested here
    PerformanceConfigConvAsm3x3U() : PerformanceConfigConvAsm3x3U(-1, -1, -1) {}

```

--- a/src/include/miopen/execution_context.hpp
+++ b/src/include/miopen/execution_context.hpp
@@ -85,10 +85,12 @@ struct ExecutionContext
     inline Handle& GetStream() const { return *stream; }
     inline void SetStream(Handle* stream_) { stream = stream_; }
 
-    ExecutionContext() = default;
     ExecutionContext(Handle* stream_) : stream(stream_) {}
-    virtual ~ExecutionContext() = default;
-    ExecutionContext(ExecutionContext const&) = default;
+    virtual ~ExecutionContext()               = default;
+    ExecutionContext(const ExecutionContext&) = default;
+    ExecutionContext(ExecutionContext&&)      = default;
+    ExecutionContext& operator=(const ExecutionContext&) = default;
+    ExecutionContext& operator=(ExecutionContext&&) = default;
 
     void DetectRocm();
 #if MIOPEN_EMBED_DB

--- a/src/include/miopen/execution_context.hpp
+++ b/src/include/miopen/execution_context.hpp
@@ -86,6 +86,8 @@ struct ExecutionContext
     inline void SetStream(Handle* stream_) { stream = stream_; }
 
     ExecutionContext(Handle* stream_) : stream(stream_) {}
+
+    ExecutionContext() {}
     virtual ~ExecutionContext()               = default;
     ExecutionContext(const ExecutionContext&) = default;
     ExecutionContext(ExecutionContext&&)      = default;

--- a/src/include/miopen/execution_context.hpp
+++ b/src/include/miopen/execution_context.hpp
@@ -88,6 +88,7 @@ struct ExecutionContext
     ExecutionContext() = default;
     ExecutionContext(Handle* stream_) : stream(stream_) {}
     virtual ~ExecutionContext() = default;
+    ExecutionContext(ExecutionContext const&) = default;
 
     void DetectRocm();
 #if MIOPEN_EMBED_DB

--- a/src/include/miopen/execution_context.hpp
+++ b/src/include/miopen/execution_context.hpp
@@ -87,12 +87,9 @@ struct ExecutionContext
 
     ExecutionContext(Handle* stream_) : stream(stream_) {}
 
-    ExecutionContext() {}
+    ExecutionContext() = default;
     virtual ~ExecutionContext()               = default;
     ExecutionContext(const ExecutionContext&) = default;
-    ExecutionContext(ExecutionContext&&)      = default;
-    ExecutionContext& operator=(const ExecutionContext&) = default;
-    ExecutionContext& operator=(ExecutionContext&&) = default;
 
     void DetectRocm();
 #if MIOPEN_EMBED_DB

--- a/src/include/miopen/execution_context.hpp
+++ b/src/include/miopen/execution_context.hpp
@@ -87,7 +87,7 @@ struct ExecutionContext
 
     ExecutionContext(Handle* stream_) : stream(stream_) {}
 
-    ExecutionContext() = default;
+    ExecutionContext()                        = default;
     virtual ~ExecutionContext()               = default;
     ExecutionContext(const ExecutionContext&) = default;
 

--- a/src/include/miopen/performance_config.hpp
+++ b/src/include/miopen/performance_config.hpp
@@ -40,8 +40,10 @@ namespace solver {
 
 struct PerfConfig
 {
-    virtual ~PerfConfig()         = default;
-    ... protected : PerfConfig()  = default;
+    virtual ~PerfConfig() = default;
+
+protected:
+    PerfConfig()                  = default;
     PerfConfig(const PerfConfig&) = default;
     PerfConfig& operator=(const PerfConfig&) = default;
 

--- a/src/include/miopen/performance_config.hpp
+++ b/src/include/miopen/performance_config.hpp
@@ -41,9 +41,7 @@ namespace solver {
 struct PerfConfig
 {
     virtual ~PerfConfig()         = default;
-    ...
-protected:
-    PerfConfig() = default;
+    ... protected : PerfConfig()  = default;
     PerfConfig(const PerfConfig&) = default;
     PerfConfig& operator=(const PerfConfig&) = default;
 

--- a/src/include/miopen/performance_config.hpp
+++ b/src/include/miopen/performance_config.hpp
@@ -42,14 +42,14 @@ struct PerfConfig
 {
     virtual ~PerfConfig() = default;
 
+    virtual void Serialize(std::ostream& stream) const = 0;
+    virtual bool Deserialize(const std::string& s)     = 0;
+    virtual std::string ToString() const;
+
 protected:
     PerfConfig()                  = default;
     PerfConfig(const PerfConfig&) = default;
     PerfConfig& operator=(const PerfConfig&) = default;
-
-    virtual void Serialize(std::ostream& stream) const = 0;
-    virtual bool Deserialize(const std::string& s)     = 0;
-    virtual std::string ToString() const;
 };
 
 std::ostream& operator<<(std::ostream& os, const PerfConfig& c);

--- a/src/include/miopen/performance_config.hpp
+++ b/src/include/miopen/performance_config.hpp
@@ -41,6 +41,7 @@ namespace solver {
 struct PerfConfig
 {
     virtual ~PerfConfig() = default;
+    PerfConfig(PerfConfig const&) = default;
 
     virtual void Serialize(std::ostream& stream) const = 0;
     virtual bool Deserialize(const std::string& s)     = 0;

--- a/src/include/miopen/performance_config.hpp
+++ b/src/include/miopen/performance_config.hpp
@@ -40,13 +40,12 @@ namespace solver {
 
 struct PerfConfig
 {
-    PerfConfig() {}
-
     virtual ~PerfConfig()         = default;
+    ...
+protected:
+    PerfConfig() = default;
     PerfConfig(const PerfConfig&) = default;
-    PerfConfig(PerfConfig&&)      = default;
     PerfConfig& operator=(const PerfConfig&) = default;
-    PerfConfig& operator=(PerfConfig&&) = default;
 
     virtual void Serialize(std::ostream& stream) const = 0;
     virtual bool Deserialize(const std::string& s)     = 0;

--- a/src/include/miopen/performance_config.hpp
+++ b/src/include/miopen/performance_config.hpp
@@ -40,8 +40,13 @@ namespace solver {
 
 struct PerfConfig
 {
-    virtual ~PerfConfig() = default;
-    PerfConfig(PerfConfig const&) = default;
+    PerfConfig() {}
+
+    virtual ~PerfConfig()         = default;
+    PerfConfig(const PerfConfig&) = default;
+    PerfConfig(PerfConfig&&)      = default;
+    PerfConfig& operator=(const PerfConfig&) = default;
+    PerfConfig& operator=(PerfConfig&&) = default;
 
     virtual void Serialize(std::ostream& stream) const = 0;
     virtual bool Deserialize(const std::string& s)     = 0;

--- a/src/include/miopen/solver.hpp
+++ b/src/include/miopen/solver.hpp
@@ -77,10 +77,6 @@ struct SolverBase
 {
     virtual ~SolverBase() = default;
 
-protected:
-    SolverBase()                  = default;
-    SolverBase(const SolverBase&) = default;
-
     /// This will retrieve the id of the solver to write to the database. By
     /// default it uses the class name. If the class is renamed, this function can
     /// overriden to keep the name to avoid DB corruption.
@@ -132,6 +128,8 @@ protected:
         static const auto result = ComputeSolverDbId(get_type_name<Solver>());
         return result;
     }
+    SolverBase()                  = default;
+    SolverBase(const SolverBase&) = default;
 
 private:
     static std::string ComputeSolverDbId(const std::string& type_name)

--- a/src/include/miopen/solver.hpp
+++ b/src/include/miopen/solver.hpp
@@ -76,6 +76,7 @@ const int wave_size = 64;
 struct SolverBase
 {
     virtual ~SolverBase() = default;
+    SolverBase(SolverBase const&) = default;
 
     /// This will retrieve the id of the solver to write to the database. By
     /// default it uses the class name. If the class is renamed, this function can

--- a/src/include/miopen/solver.hpp
+++ b/src/include/miopen/solver.hpp
@@ -75,13 +75,11 @@ const int wave_size = 64;
 /// There could be multiple solvers of the same algorithm for a problem config.
 struct SolverBase
 {
-    SolverBase() {}
-
-    virtual ~SolverBase()         = default;
+    virtual ~SolverBase() = default;
+    ...
+protected:
+    SolverBase() = default;
     SolverBase(const SolverBase&) = default;
-    SolverBase(SolverBase&&)      = default;
-    SolverBase& operator=(const SolverBase&) = default;
-    SolverBase& operator=(SolverBase&&) = default;
 
     /// This will retrieve the id of the solver to write to the database. By
     /// default it uses the class name. If the class is renamed, this function can

--- a/src/include/miopen/solver.hpp
+++ b/src/include/miopen/solver.hpp
@@ -75,8 +75,10 @@ const int wave_size = 64;
 /// There could be multiple solvers of the same algorithm for a problem config.
 struct SolverBase
 {
-    virtual ~SolverBase()         = default;
-    ... protected : SolverBase()  = default;
+    virtual ~SolverBase() = default;
+
+protected:
+    SolverBase()                  = default;
     SolverBase(const SolverBase&) = default;
 
     /// This will retrieve the id of the solver to write to the database. By

--- a/src/include/miopen/solver.hpp
+++ b/src/include/miopen/solver.hpp
@@ -75,10 +75,8 @@ const int wave_size = 64;
 /// There could be multiple solvers of the same algorithm for a problem config.
 struct SolverBase
 {
-    virtual ~SolverBase() = default;
-    ...
-protected:
-    SolverBase() = default;
+    virtual ~SolverBase()         = default;
+    ... protected : SolverBase()  = default;
     SolverBase(const SolverBase&) = default;
 
     /// This will retrieve the id of the solver to write to the database. By

--- a/src/include/miopen/solver.hpp
+++ b/src/include/miopen/solver.hpp
@@ -75,8 +75,13 @@ const int wave_size = 64;
 /// There could be multiple solvers of the same algorithm for a problem config.
 struct SolverBase
 {
-    virtual ~SolverBase() = default;
-    SolverBase(SolverBase const&) = default;
+    SolverBase() {}
+
+    virtual ~SolverBase()         = default;
+    SolverBase(const SolverBase&) = default;
+    SolverBase(SolverBase&&)      = default;
+    SolverBase& operator=(const SolverBase&) = default;
+    SolverBase& operator=(SolverBase&&) = default;
 
     /// This will retrieve the id of the solver to write to the database. By
     /// default it uses the class name. If the class is renamed, this function can

--- a/src/include/miopen/utility/transposing_solver.hpp
+++ b/src/include/miopen/utility/transposing_solver.hpp
@@ -77,15 +77,11 @@ struct TransposeInvokeParams : InvokeParams
 
 struct TransposePseudoSolver
 {
-    TransposePseudoSolver() {}
     virtual ~TransposePseudoSolver()                    = default;
+    ...
+protected:
+    TransposePseudoSolver() = default;
     TransposePseudoSolver(const TransposePseudoSolver&) = default;
-    TransposePseudoSolver(TransposePseudoSolver&&)      = default;
-    TransposePseudoSolver& operator=(const TransposePseudoSolver&) = default;
-    TransposePseudoSolver& operator=(TransposePseudoSolver&&)               = default;
-    virtual std::string GetTranspose() const                                = 0;
-    virtual ConvSolution GetSolution(const ExecutionContext& ctx,
-                                     const TransposeProblem& problem) const = 0;
 };
 
 template <class Derived, class Interface>

--- a/src/include/miopen/utility/transposing_solver.hpp
+++ b/src/include/miopen/utility/transposing_solver.hpp
@@ -78,8 +78,10 @@ struct TransposeInvokeParams : InvokeParams
 struct TransposePseudoSolver
 {
     virtual ~TransposePseudoSolver()                    = default;
-    ... protected : TransposePseudoSolver()             = default;
     TransposePseudoSolver(const TransposePseudoSolver&) = default;
+
+protected:
+    TransposePseudoSolver() = default;
 };
 
 template <class Derived, class Interface>

--- a/src/include/miopen/utility/transposing_solver.hpp
+++ b/src/include/miopen/utility/transposing_solver.hpp
@@ -78,9 +78,7 @@ struct TransposeInvokeParams : InvokeParams
 struct TransposePseudoSolver
 {
     virtual ~TransposePseudoSolver()                    = default;
-    ...
-protected:
-    TransposePseudoSolver() = default;
+    ... protected : TransposePseudoSolver()             = default;
     TransposePseudoSolver(const TransposePseudoSolver&) = default;
 };
 

--- a/src/include/miopen/utility/transposing_solver.hpp
+++ b/src/include/miopen/utility/transposing_solver.hpp
@@ -78,6 +78,7 @@ struct TransposeInvokeParams : InvokeParams
 struct TransposePseudoSolver
 {
     virtual ~TransposePseudoSolver()                                        = default;
+    TransposePseudoSolver(TransposePseudoSolver const&) = default;
     virtual std::string GetTranspose() const                                = 0;
     virtual ConvSolution GetSolution(const ExecutionContext& ctx,
                                      const TransposeProblem& problem) const = 0;

--- a/src/include/miopen/utility/transposing_solver.hpp
+++ b/src/include/miopen/utility/transposing_solver.hpp
@@ -77,8 +77,12 @@ struct TransposeInvokeParams : InvokeParams
 
 struct TransposePseudoSolver
 {
-    virtual ~TransposePseudoSolver()                                        = default;
-    TransposePseudoSolver(TransposePseudoSolver const&) = default;
+    TransposePseudoSolver() {}
+    virtual ~TransposePseudoSolver()                    = default;
+    TransposePseudoSolver(const TransposePseudoSolver&) = default;
+    TransposePseudoSolver(TransposePseudoSolver&&)      = default;
+    TransposePseudoSolver& operator=(const TransposePseudoSolver&) = default;
+    TransposePseudoSolver& operator=(TransposePseudoSolver&&)               = default;
     virtual std::string GetTranspose() const                                = 0;
     virtual ConvSolution GetSolution(const ExecutionContext& ctx,
                                      const TransposeProblem& problem) const = 0;

--- a/src/include/miopen/utility/transposing_solver.hpp
+++ b/src/include/miopen/utility/transposing_solver.hpp
@@ -77,11 +77,14 @@ struct TransposeInvokeParams : InvokeParams
 
 struct TransposePseudoSolver
 {
-    virtual ~TransposePseudoSolver()                    = default;
-    TransposePseudoSolver(const TransposePseudoSolver&) = default;
+    virtual ~TransposePseudoSolver()                                        = default;
+    virtual std::string GetTranspose() const                                = 0;
+    virtual ConvSolution GetSolution(const ExecutionContext& ctx,
+                                     const TransposeProblem& problem) const = 0;
 
 protected:
-    TransposePseudoSolver() = default;
+    TransposePseudoSolver()                             = default;
+    TransposePseudoSolver(const TransposePseudoSolver&) = default;
 };
 
 template <class Derived, class Interface>

